### DISCUSSION TITLE
Emscripten (wasm32) support

### DISCUSSION THREAD

**Discussion by ethanaobrien** [OpenRCT2/OpenRCT2#23392] [Feature requests]
Webassembly - initially added by @janisozaur in #6254  a total of 7 years ago now (wow!) Time sure flies

However, adding support for a (relatively) obscure target, with no build documentation or tests, is bound to become broken, fast!

I'd like for that to change, though sadly I am not familiar enough with this project, and am not fluent enough in C++ to do this cleanly.

Over the course of several days (and several hours per day) banging my head against the wall (not literally), figuring out why cmake cant find zlib, trying to figure out why [this was ever a thing](https://github.com/olydis/OpenRCT2/blob/0a831dc8a61f58ed3c8b466952400d852195c7d9/www/index.html#L2581), etc etc etc... I did finally manage to get the latest version of OpenRCT2 to run well within emscripten! (Big thanks to @mstrodl in [this repo](https://github.com/Mstrodl/ORCT2-web) for helping me figure out the cmake side!)

However, 
There were several parts where it's obvious I had no clue what I was doing. Specifically, ~~several times in [String.cpp](https://github.com/ethanaobrien/OpenRCT2/blob/11c34fe434f50df0e41bbd5e08c9a6ce10e035ed/src/openrct2/core/String.cpp#L61) where I would just do what was needed to make emscripten happy~~ (this was fixed) ~~([Like ignoring an error](https://github.com/ethanaobrien/OpenRCT2/blob/11c34fe434f50df0e41bbd5e08c9a6ce10e035ed/src/openrct2/entity/Peep.cpp#L2743), probably caused by some sort of race-condition somewhere else. I dont know the codebase well enough to debug this)~~ this also magically got fixed

However, I think it's decently-ish well done (in the sense that it works).

My request:
I'd like if someone familiar with the codebase of this project would give their opinions and help fix-up on [my changes](https://github.com/OpenRCT2/OpenRCT2/compare/develop...ethanaobrien:OpenRCT2:develop), so we can get this upstreamed!

My changes include a static deployment of a page that plays the game. The main difference with how I aimed to implement it, is I do not store the RCT2 data server-side. The user must upload a ZIP file containing the RCT2 data files. I've also added the ability for the user to import/export their save data. This way it could be hosted on a public official instance

I'm hesitant to open a PR, because I know I likely won't be able to fix my code up to the standard, I'd like to get some new eyes first!

Please try it out here: https://rct2-test.netlify.app

Edits: I spent the entire day cleaning up the build target and patches, squashed my commits, and now I think it looks and works much better!

**Comment by Mstrodl** [comment]
Awesome! Would be cool to see my http support get added too so you can get stuff like the changelog: https://github.com/Mstrodl/OpenRCT2/blob/e0b4a63789012c987e96d22b33375699704b9bd8/src/openrct2/core/Http.Emscripten.cpp

**Reply by ethanaobrien** [reply]
I do have this planned! I was hoping to get the base emscripten support upstreamed, and slowly add more to it to not overwhelm the reviewers or myself. Thank you for your help!

**Comment by janisozaur** [comment]
I like the changes and I think they are concise enough they get a chance to be upstreamed.

My comments so far:
1. you claim you want the user to upload OpenRCT2 data, but what you really want is RCT2 assets. 
2. The default display engine should be changed to software+hw display
3. Navigating with RMB feels awkward, there is some scaling applied that is different from the render scale
4. The process for building should be automated, there should be a new docket image created with all the necessary toolchains and libraries installed, that would be used to build in CI.
5. With this now built, we should be able to deploy it automatically somewhere (GitHub static pages or our digitalocean server) for users to consume. 

Let me know how you find these and if you want to fix them or should we take over some of the points.

**Reply by Mstrodl** [reply]
> Navigating with RMB feels awkward, there is some scaling applied that is different from the render scale

This is fixed in my fork. I forget what I had to do, but it basically is because of high-dpi displays / display scaling

**Reply by Mstrodl** [reply]
Here it is: https://github.com/Mstrodl/OpenRCT2/commit/e0b4a63789012c987e96d22b33375699704b9bd8#diff-c511011e26795ba92b6780fe6e8db10a912d6014fd05c5789e6849383b48cb5f

**Reply by Mstrodl** [reply]
Despite what it looks like, it will still correctly display the game in high-dpi. I'm guessing this is a bug or incorrectly supported feature in the emscripten SDL port. I spent some time trying to figure it out but wasn't able to make a ton of progress.

**Reply by ethanaobrien** [reply]
Hi @janisozaur, thank you for looking at it. I've done the following (https://github.com/ethanaobrien/OpenRCT2/commit/468643d1a9e3f71c96429e172963d92af6c0c69a):

1. Fixed
2. This has been changed!
3. That is something I will need help with, with RMB disabled, the mouse is incredibly sensitive to the point moving using the mouse isn't useful:

https://github.com/user-attachments/assets/f2388300-c017-4402-8ce6-de8b79499745



5. I've added a dockerfile you can use to build with `docker build -f emscripten/Dockerfile --output type=local,dest=./emscripten-output .`
